// circle-blvd/payment.js
// payment.js
var guard = require('circle-blvd/errors').guard;

var stripeProcessor = require('stripe');
var stripe = undefined;

module.exports = function (db) {
	
	var setApiKey = function (secretKey) {
		stripe = stripeProcessor(secretKey);
	};

	var donate = function (stripeTokenId, stripeAmount, callback) {
		var donation = {
			amount: stripeAmount,
			currency: "usd",
			card: stripeTokenId,
			description: "Donation",
			statement_description: "Donation"
		};

		stripe.charges.create(donation, callback);
	};

	var subscribe = function (user, stripeTokenId, planName, callback) {
		var planId = undefined;
		
		if (planName === 'Supporter') {
			planId = '2014-06-supporter';
		}
		if (planName === 'Organizer') {
			planId = '2014-06-organizer';
		}
		if (planName === 'Patron') {
			planId = '2014-06-patron';
		}
		if (planName === 'Thirty') {
			planId = '2016-04-thirty';
		}

		if (planName === 'Ten') {
			planId = '2016-04-ten';
		}

		if (!planId) {
			var error = new Error("Invalid plan name");
			error.status = 400;
			return callback(error);
		}

		var onSuccess = function (updatedUser) {
			callback(null, updatedUser.subscription);
		};
		var onError = function (err) {
			// TODO: Technically it's possible to update
			// the Stripe data and not update our own 
			// data, so we should have a fall-back plan
			// if that happens.
			callback(err);
		};

		if (!user.subscription) {
			// First time here!
			var newCustomer = {
				description: user.name + " (" + user.id + ")",
				card: stripeTokenId,
				plan: planId,
				metadata: {
					"id": user.id
				}
			};

			stripe.customers.create(newCustomer, guard(callback, function (customer) {
				var sub = {};
				sub.created = customer.created;
				sub.customerId = customer.id;
				sub.subscriptionId = customer.subscriptions.data[0].id;
				sub.planName = planName;

				user.subscription = sub;
				db.users.update(user, onSuccess, onError);
			}));
		}
		else {
			// Returning customer
			var customerId = user.subscription.customerId;
			var subscriptionId = user.subscription.subscriptionId;
			var newPlan = {
				card: stripeTokenId,
				plan: planId
			};
			stripe.customers.updateSubscription(
				customerId, subscriptionId, newPlan,
				guard(callback, function (subscription) {
					var newSub = user.subscription;
					newSub.subscriptionId = subscription.id;
					newSub.planName = planName;
					newSub.updated = subscription.start;

					user.subscription = newSub;
					db.users.update(user, onSuccess, onError);
				})
			);
		}
	};

	var unsubscribe = function (user, callback) {
		// Just delete the Stripe customer, since they
		// only have one subscription anyway.
		var customerId = user.subscription.customerId;
		stripe.customers.del(customerId, guard(callback, function (confirm) {

			var onSuccess = function (updatedUser) {
				callback(null, updatedUser.subscription);
			};
			var onError = function (err) {
				// TODO: Technically it's possible to update
				// the Stripe data and not update our own 
				// data, so we should have a fall-back plan
				// if that happens.
				callback(err);
			};

			user.subscription = null;
			db.users.update(user, onSuccess, onError);
		}));
	};

	return {
		donate: donate,
		subscribe: subscribe,
		unsubscribe: unsubscribe,
		setApiKey: setApiKey
	};
};
