// circle-blvd/session.js
// session.js
//
var expressSession    = require('express-session');
var couchSessionStore = require('circle-blvd/couch-session-store');
var uidSafe           = require('uid-safe');

var getCookieSettings = function () {
    // TODO: Check settings to guess if https is running.
    // Or actually figure out if https is running, and if so
    // use secure cookies
    var oneHour = 3600000;
    var oneWeek = 7 * 24 * oneHour;
    var sixWeeks = 6 * oneWeek;
    var cookieSettings = {
        path: '/',
        httpOnly: true,
        secure: false,
        maxAge: sixWeeks
    };

    return cookieSettings;
};

var middleware = function (sessionSecret, database) {
    var SessionStore = couchSessionStore(expressSession, database);
    var cookieSettings = getCookieSettings();
    var sessionMiddleware = expressSession({ 
        store: new SessionStore(),
        secret: sessionSecret,
        cookie: cookieSettings,
        // TODO: We might want resave to be false
        // More info: https://github.com/expressjs/session#options
        resave: true,
        saveUninitialized: true,
        // Do not allow session IDs to start with an underscore. 
        // CouchDB does not allow us to use document IDs that start with _.
        genid: function (req) {
            var id = undefined;
            while (!id) {
                id = uidSafe.sync(24);
                if (id[0] === "_") {
                    id = undefined;
                }
            }
            return id;
        }
    });

    return sessionMiddleware;
};

exports.middleware = middleware;
